src/utils/changelog.ts

```diff
@@ -0,0 +1,396 @@
+  markdown: string,
+  changeset: Changeset
+): string {
+  // Try to locate the top-most non-empty header, no matter what is inside
+  const { start, padding } = locateChangeset(markdown, Boolean) || {
+    padding: '',
+  };
+  const body = changeset.body || `${padding}${DEFAULT_CHANGESET_BODY}`;
+  let header;
+  if (start?.[3]) {
+    const underline = new Array(changeset.name.length + 1).join('-');
+    header = `${changeset.name}\n${underline}`;
+  } else {
+    header = `## ${changeset.name}`;
+  }
+  const newSection = `${padding}${header}\n\n${body.replace(
+    /^/gm,
+    padding
+  )}\n\n`;
+  const startIdx = start?.index ?? markdown.length;
+
+  return markdown.slice(0, startIdx) + newSection + markdown.slice(startIdx);
+}
+
+interface Commit {
+  hash: string;
+  title: string;
+  hasPRinTitle: boolean;
+  pr: string | null;
+  milestone: string | null;
+}
+
+interface Milestone {
+  title?: string;
+  description?: string;
+  state?: 'OPEN' | 'CLOSED';
+}
+type MilestoneWithPRs = Milestone & {
+  prs: string[];
+};
+
+// This is set to 8 since GitHub and GitLab prefer that over the default 7 to
+// avoid collisions.
+const SHORT_SHA_LENGTH = 8;
+function formatCommit(commit: Commit): string {
+  let text = `- ${commit.title}`;
+  if (!commit.hasPRinTitle) {
+    const link = commit.pr
+      ? `#${commit.pr}`
+      : commit.hash.slice(0, SHORT_SHA_LENGTH);
+    text = `${text} (${link})`;
+  }
+  return text;
+}
+
+export async function generateChangesetFromGit(
+  git: SimpleGit,
+  rev: string
+): Promise<string> {
+  const gitCommits = await getChangesSince(git, rev);
+
+  const githubCommits = await getMilestoneAndPRFromCommits(
+    gitCommits.map(commit => commit.hash)
+  );
+
+  const milestones: Record</*milestone #*/ string, MilestoneWithPRs> = {};
+  const commits: Record</*hash*/ string, Commit> = {};
+  const leftovers: Commit[] = [];
+  const missing: Commit[] = [];
+  for (const gitCommit of gitCommits) {
+    const hash = gitCommit.hash;
+
+    const githubCommit = githubCommits[hash];
+    const commit = (commits[hash] = {
+      hash: hash,
+      title: gitCommit.message,
+      hasPRinTitle: Boolean(gitCommit.pr),
+      ...githubCommit,
+    });
+
+    if (!githubCommit) {
+      missing.push(commit);
+    }
+    if (!commit.milestone) {
+      leftovers.push(commit);
+    } else {
+      const milestone = milestones[commit.milestone] || { prs: [] };
+      // We _know_ the PR exists as milestones are attached to PRs
+      milestone.prs.push(commit.pr as string);
+      milestones[commit.milestone] = milestone;
+    }
+  }
+
+  if (missing.length > 0) {
+    logger.warn(
+      `The following commits were not found on GitHub:`,
+      missing.map(commit => `${commit.hash.slice(0, 8)} ${commit.title}`)
+    );
+  }
+
+  const milestonesInfo = await getMilestonesDetails(Object.keys(milestones));
+
+  const changelogSections = [];
+  for (const milestoneNum of Object.keys(milestones)) {
```

Comment:
What does the sort end up being on this? Comment? Test?

Reply:
No sort at the moment, don't see a strong need yet. JS dictionaries now implicit sorting based on the insertion order so this should always produce consistent results.

Reply:
Sgtm. Also, excellent footnote game. 😁